src/libshadertrap/src/executor.cc

```diff
@@ -1201,41 +1203,171 @@ bool Executor::CheckEqualBuffers(CommandAssertEqual* assert_equal) {
-                buffers[index]);
-    mapped_buffer[index] =
-        static_cast<uint8_t*>(gl_functions_->glMapBufferRange_(
-            GL_ARRAY_BUFFER, 0, static_cast<GLsizeiptr>(buffer_size[index]),
-            GL_MAP_READ_BIT));
-    if (mapped_buffer[index] == nullptr) {
-      // TODO(afd): If index == 1 should we unmap buffers[0] before
-      //  returning? Or are we OK so long as we eventually destroy that buffer?
-      GL_CHECKERR(&assert_equal->GetStartToken(), "glMapBufferRange");
-      return false;
-    }
+  bool result = true;
+  size_t start_index = 0;
+  std::vector<CommandAssertEqual::FormatEntry>& format_entries = assert_equal->GetFormatEntries();
+  const Token t = assert_equal->GetStartToken();
+  std::unique_ptr<Token> tmp(std::move(new Token(t)));
+  if(format_entries.size() == 0){
+     format_entries.push_back(
+              {std::move(tmp), CommandAssertEqual::FormatEntry::Kind::kByte, (size_t)buffer_size[0]});
   }
 
-  bool result = true;
-  for (size_t index = 0; index < static_cast<size_t>(buffer_size[0]); index++) {
-    // We only get here if the calls to glMapBufferRange succeeded, in which
-    // case the contents of |mapped_buffer| cannot be null.
-    uint8_t value_1 =
-        mapped_buffer[0][index];  // NOLINT(clang-analyzer-core.NullDereference)
-    uint8_t value_2 =
-        mapped_buffer[1][index];  // NOLINT(clang-analyzer-core.NullDereference)
-    if (value_1 != value_2) {
-      std::stringstream stringstream;
-      stringstream << "Byte mismatch at index " << index << ": "
-                   << assert_equal->GetArgumentIdentifier1() << "[" << index
-                   << "] == " << static_cast<uint32_t>(value_1) << ", "
-                   << assert_equal->GetArgumentIdentifier2() << "[" << index
-                   << "] == " << static_cast<uint32_t>(value_2);
-      message_consumer_->Message(MessageConsumer::Severity::kError,
-                                 &assert_equal->GetStartToken(),
-                                 stringstream.str());
-      result = false;
+  for(auto& format_entry: format_entries){
+    switch(format_entry.kind){
+      case CommandAssertEqual::FormatEntry::Kind::kSkip:
+          start_index += format_entry.count;
+          break;
+      case CommandAssertEqual::FormatEntry::Kind::kByte:
+      {
+        uint8_t* mapped_buffer[2]{nullptr, nullptr};
+        for (auto index : {0, 1}) {
+          GL_SAFECALL(&assert_equal->GetStartToken(), glBindBuffer, GL_ARRAY_BUFFER,
+                      buffers[index]);
+          mapped_buffer[index] =
+              static_cast<uint8_t*>(gl_functions_->glMapBufferRange_(
+                  GL_ARRAY_BUFFER, 0, static_cast<GLsizeiptr>(buffer_size[index]),
+                  GL_MAP_READ_BIT));
+          if (mapped_buffer[index] == nullptr) {
+            GL_CHECKERR(&assert_equal->GetStartToken(), "glMapBufferRange");
+            return false;
+          }
```

Comment:
It looks like you are doing doing copies of the buffer again and again for every different format specifier.

That's not good: there's a large amount of duplicated code, and it involves a lot of wasted work.

Instead, revert to the existing approach where before the main `for` loop, the two buffers are copied to host memory in their entirity.

Then you can access the mapped buffers at the indices that you require when looping through the format entries.